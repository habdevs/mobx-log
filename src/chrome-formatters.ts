import { isObservableArray, isObservableObject, toJS } from 'mobx';

type Css = { style: string };
type Chunk = [string, Css, string];
type Header = [string, Chunk, [string, string]];
type Body = [string, Css, ...any[]];

export type ChromeFormatter<T> = {
  header: (argument: T) => Header | null;
  hasBody: (argument: T) => boolean;
  body: (argument: T) => Body;
};

const styles = {
  mobxName: { style: 'color: rgb(232,98,0)' },
  nullName: { style: 'color: #777' },
  list: {
    style:
      'list-style-type: none; padding: 0; margin: 0 0 0 12px; font-style: normal',
  },
};

const reference = (object: unknown) => {
  if (typeof object === 'undefined' || typeof object === null) {
    return ['span', styles.nullName, typeof object];
  }
  return ['object', { object }];
};

const renderIterableHeader = (
  iterable: ArrayLike<unknown>,
  name: string
): Header => [
  'span',
  ['span', styles.mobxName, name],
  ['span', `[${iterable.length}]`],
];

const hasBody = (collection: ArrayLike<unknown>) => collection.length > 0;

const renderIterableBody = (
  collection: ArrayLike<unknown> | {},
  mapper: (val: [string, unknown]) => unknown
): Body => {
  const children = Object.entries(toJS(collection)).map(mapper);
  return ['ol', styles.list, ...children];
};

export class ObjectFormatter implements ChromeFormatter<{}> {
  header(argument: {}): Header | null {
    return isObservableObject(argument)
      ? renderIterableHeader(Object.keys(argument), 'Object')
      : null;
  }

  hasBody(argument: {}): boolean {
    return hasBody(Object.keys(argument));
  }

  body(argument: {}) {
    return renderIterableBody(argument, ([key, value]) => [
      'li',
      {},
      reference(key),
      ': ',
      reference(value),
    ]);
  }
}

export class ArrayFormatter implements ChromeFormatter<ArrayLike<unknown>> {
  body(argument: ArrayLike<unknown>): Body {
    return renderIterableBody(argument, ([_, value]) => [
      'li',
      reference(value),
    ]);
  }

  hasBody(argument: ArrayLike<unknown>): boolean {
    return hasBody(argument);
  }

  header(argument: ArrayLike<unknown>): Header | null {
    return isObservableArray(argument)
      ? renderIterableHeader(argument, 'Array')
      : null;
  }
}
