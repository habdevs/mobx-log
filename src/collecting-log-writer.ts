import { LogWriter } from './types';

export class CollectingLogWriter implements LogWriter {
  history: string[] = [];

  write(...messages: unknown[]) {
    const strings = messages.map((message): string =>
      (message as any).toString()
    );
    this.history.push(strings.join(' '));
  }
}
