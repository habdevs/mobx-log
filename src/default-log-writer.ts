import { LogWriter } from './types';

export class DefaultLogWriter implements LogWriter {
  write(...messages: unknown[]) {
    console.log(...messages);
  }
}
