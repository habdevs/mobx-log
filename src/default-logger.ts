import { ActionEvent, ComputedEvent, Logger, ObservableEvent } from './types';
import { LogWriter } from './log-writer';

const colors = {
  red: '#FF6157',
  blue: '#28B7F6',
  orange: '#FFBC2D',
  green: '#2BC941',
};

const buildStyles = (color: string) => `color:${color}`;

export class DefaultLogger implements Logger {
  constructor(private logWriter: LogWriter) {}

  logObservable(event: ObservableEvent) {
    const info = [
      '%c[O]',
      buildStyles(colors.orange),
      event.name,
      event.oldValue,
      '->',
      event.newValue,
    ];

    this.logWriter.write(...info);
  }

  logAction(event: ActionEvent) {
    const info = [
      '%c[A]',
      buildStyles(colors.red),
      event.name,
      event.arguments,
    ];

    this.logWriter.write(...info);
  }

  logComputed(event: ComputedEvent) {
    const info = [
      '%c[C]',
      buildStyles(colors.blue),
      event.name,
      event.oldValue,
      '->',
      event.newValue,
    ];

    this.logWriter.write(...info);
  }
}
