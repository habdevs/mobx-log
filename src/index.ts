export { enableLogging } from "./enable-logging";
