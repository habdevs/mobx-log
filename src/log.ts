import type { PureSpyEvent } from "mobx/src/core/spy";
import { IEqualsComparer } from "mobx";
import { Logger } from "./logger";

type SpyEvent = PureSpyEvent & {
  object: any;
  name: string;
};

export type LogOptions = {
  logger: Logger;
  compare: IEqualsComparer<unknown>;
};

export const log = (event: SpyEvent, options: LogOptions): void => {
  const { logger, compare } = options;

  if (event.type === "update") {
    if (event.observableKind === "computed") {
      if (!compare(event.oldValue, event.newValue)) {
        logger.logComputed({
          name: event.debugObjectName,
          oldValue: event.oldValue,
          newValue: event.newValue,
        });
      }
    }
    if (event.observableKind === "object") {
      logger.logObservable({
        name: `${event.debugObjectName}.${event.name}`,
        newValue: event.newValue,
        oldValue: event.oldValue,
      });
    }
  }
  if (event.type === "action") {
    logger.logAction({
      name: `${event.object.constructor.name}.${event.name}`,
      arguments: event.arguments,
    });
  }
};
