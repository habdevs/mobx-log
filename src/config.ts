import { DefaultLogger } from './default-logger';
import { Logger } from './types';
import { DefaultLogWriter } from './log-writer';

type Config = {
  logger: Logger;
};

export const config: Config = {
  logger: new DefaultLogger(new DefaultLogWriter()),
};

export const configureMakeLoggable = (options: Config) => {
  config.logger = options.logger;
};
