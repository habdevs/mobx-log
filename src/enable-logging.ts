import { spy } from "mobx";
import { log, LogOptions } from "./log";
import { DefaultLogger } from "./default-logger";
import { now } from "./now";

const defaultOptions = {
  logger: new DefaultLogger(console.log, now),
};

export const enableLogging = (options: LogOptions = defaultOptions) => {
  spy((event) => log(event, options));
};
