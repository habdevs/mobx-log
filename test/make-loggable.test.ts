import { autorun, makeAutoObservable } from 'mobx';
import { DefaultLogger } from '../src/default-logger';
import { CollectingLogWriter } from '../src/collecting-log-writer';
import { configureMakeLoggable, makeLoggable } from '../src';

const collectingWriter = new CollectingLogWriter();

configureMakeLoggable({
  logger: new DefaultLogger(collectingWriter, () => '13:13:13'),
});

class StoreWithComputed {
  value = 0;

  constructor() {
    makeAutoObservable(this);
    makeLoggable(this);
  }

  increment() {
    this.value++;
  }

  get isDividedBy3() {
    return this.value % 3 === 0;
  }
}

class StoreOnlyObservables {
  age?: number;
  name?: string;

  constructor() {
    makeAutoObservable(this);
    makeLoggable(this);
  }

  init() {
    this.age = 24;
    this.name = 'Test';
  }
}

class StoreWithoutLog {
  counter = 1;

  constructor() {
    makeAutoObservable(this);
  }

  increment() {
    this.counter++;
  }

  get isEven() {
    return this.counter % 2 === 0;
  }
}

describe('makeLoggable', () => {
  it('logs', async (done) => {
    const storeWithComputed = new StoreWithComputed();
    const storeOnlyObservables = new StoreOnlyObservables();
    const storeWithoutLog = new StoreWithoutLog();

    let isDividedBy3 = false;
    let isEven = false;
    autorun(() => {
      isDividedBy3 = storeWithComputed.isDividedBy3;
      isEven = storeWithoutLog.isEven;
    });

    storeWithoutLog.increment();
    storeWithComputed.increment();
    storeOnlyObservables.init();
    expect(isDividedBy3).toBeFalsy();
    expect(isEven).toBeTruthy();
    expect(storeWithComputed.value).toBe(1);
    expect(collectingWriter.history).toMatchSnapshot();

    storeWithComputed.increment();
    storeWithComputed.increment();

    expect(collectingWriter.history).toMatchSnapshot();

    done();
  });
});
