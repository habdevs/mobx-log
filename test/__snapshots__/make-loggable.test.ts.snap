// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`makeLoggable logs 1`] = `
Array [
  "%c[C] color:#28B7F6 StoreWithComputed.isDividedBy3 [object Object] -> true",
  "%c[A] color:#FF6157 StoreWithComputed.increment ",
  "%c[O] color:#FFBC2D StoreWithComputed.value 0 -> 1",
  "%c[C] color:#28B7F6 StoreWithComputed.isDividedBy3 true -> false",
  "%c[A] color:#FF6157 StoreOnlyObservables.init ",
]
`;

exports[`makeLoggable logs 2`] = `
Array [
  "%c[C] color:#28B7F6 StoreWithComputed.isDividedBy3 [object Object] -> true",
  "%c[A] color:#FF6157 StoreWithComputed.increment ",
  "%c[O] color:#FFBC2D StoreWithComputed.value 0 -> 1",
  "%c[C] color:#28B7F6 StoreWithComputed.isDividedBy3 true -> false",
  "%c[A] color:#FF6157 StoreOnlyObservables.init ",
  "%c[A] color:#FF6157 StoreWithComputed.increment ",
  "%c[O] color:#FFBC2D StoreWithComputed.value 1 -> 2",
  "%c[A] color:#FF6157 StoreWithComputed.increment ",
  "%c[O] color:#FFBC2D StoreWithComputed.value 2 -> 3",
  "%c[C] color:#28B7F6 StoreWithComputed.isDividedBy3 false -> true",
]
`;
